server/src/main/java/io/crate/planner/operators/CorrelatedJoin.java

```diff
@@ -123,15 +113,17 @@ public ExecutionPlan build(DependencyCarrier executor,
             subQueryResults,
             params,
             inputPlan.outputs(),
-            outputs
+            outputs()
         );
         sourcePlan.addProjection(projection);
         return sourcePlan;
     }
 
     @Override
     public List<Symbol> outputs() {
-        return outputs;
+        LinkedHashSet<Symbol> result = new LinkedHashSet<>(inputPlan.outputs());
+        result.add(selectSymbol);
+        return List.copyOf(result);
```

Comment:
Could this return `Lists.concat(inputPlan.outputs(), selectSymbol)` ?

Would be cheaper and it shouldn't be the responsibility of `CorrelatedJoin` to de-duplicate it's childs outputs.

Reply:
Also - if this works - wouldn't it be possible to adapt the `pruneOutputsExcept` logic like this?

```
diff --git a/server/src/main/java/io/crate/planner/operators/CorrelatedJoin.java b/server/src/main/java/io/crate/planner/operators/CorrelatedJoin.java
index 42b1a4ba7ab..f3988105ec4 100644
--- a/server/src/main/java/io/crate/planner/operators/CorrelatedJoin.java
+++ b/server/src/main/java/io/crate/planner/operators/CorrelatedJoin.java
@@ -19,11 +19,10 @@
  * software solely pursuant to the terms of the relevant commercial agreement.
  */
 
 package io.crate.planner.operators;
 
-import java.util.ArrayList;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.SequencedCollection;
 import java.util.Set;
@@ -161,22 +160,15 @@ public Void visitOuterColumn(OuterColumn outerColumn, Void ignored) {
         LogicalPlan newInputPlan = inputPlan.pruneOutputsExcept(toCollect);
 
         if (inputPlan == newInputPlan) {
             return this;
         }
-        List<Symbol> newOutputs = new ArrayList<>();
-        for (Symbol output: outputs) {
-            if (toCollect.contains(output)) {
-                newOutputs.add(output);
-            }
-        }
-
         return new CorrelatedJoin(
             newInputPlan,
             selectSymbol,
             subQueryPlan,
-            newOutputs
+            Lists.concat(newInputPlan.outputs(), selectSymbol)
         );
     }
 
     @Override
     public Map<LogicalPlan, SelectSymbol> dependencies() {
```

Afaik `outputs` can be called many times during the planner/optimizer phase. Might have some cost to run this on each call.

Reply:
thx!